core/src/Streamly/Internal/FileSystem/Path.hs

```diff
@@ -0,0 +1,485 @@
+-- as null terminated.
+-- XXX Maintain the Array with null termination because Unix system calls
+-- require a null terminated string, also they return null terminated strings
+-- as paths. Implementation of path append will have to handle the null
+-- termination. Or we can choose to always copy the array when using it in
+-- system calls.
+
+-- XXX The eq instance needs to make sure that the paths are equivalent. If we
+-- normalize the paths we can do a byte comparison. However, on windows paths
+-- are case insensitive but the case is preserved, therefore, we cannot
+-- normalize and need to do case insensitive comparison.
+
+------------------------------------------------------------------------------
+-- Types
+------------------------------------------------------------------------------
+
+-- | A type representing file system paths for directories or files.
+newtype Path = Path (Array WORD_TYPE) -- deriving Eq
+
+-- XXX Do we need a type for file or dir Name as names cannot have the
+-- separator char and there may be other restrictions on names? For example,
+-- length restriction.  A file name cannot be "." or "..". We can use the types
+-- "File Name" and "Dir Name" to represent names. Also, file systems may put
+-- limits on names. Can have an IsName type class with members Name, (File
+-- Name), (Dir Name).
+
+-- | A type representing a file path.
+newtype File a = File a
+
+-- | A type representing a directory path.
+newtype Dir a = Dir a
+
+-- | A type representing absolute paths.
+newtype Abs a = Abs a
+
+-- | A type representing relative paths.
+newtype Rel a = Rel a
+
+------------------------------------------------------------------------------
+-- Conversions
+------------------------------------------------------------------------------
+
+-- | A member of 'IsPath' knows how to convert to and from the 'Path' type.
+class IsPath a where
+    -- | Like 'fromPath' but does not check the properties of 'Path'. Provides
+    -- performance and simplicity when we know that the properties of the path
+    -- are already verified, for example, when we get the path from the file
+    -- system or the OS APIs.
+    fromPathUnsafe :: Path -> a
+
+    -- XXX Use MonadThrow instead of Maybe?
+
+    -- | Convert a raw 'Path' to other forms of well-typed paths. It may fail
+    -- if the path does not satisfy the properties of the target type.
+    --
+    -- Path components may have limits.
+    -- Total path length may have a limit.
+    fromPath :: MonadThrow m => Path -> m a
+
+    -- | Convert a well-typed path to a raw 'Path'. Never fails.
+    toPath :: a -> Path
+
+instance IsPath Path where
+    fromPathUnsafe = id
+    fromPath = pure
+    toPath = id
+
+instance IsPath (File Path) where
+    fromPathUnsafe p = File p
+    fromPath p = pure (File p)
+    toPath (File p) = p
+
+instance IsPath (Dir Path) where
+    fromPathUnsafe p = Dir p
+    fromPath p = pure (Dir p)
```

Comment:
We don't check anything here? Is it a TODO?

Reply:
Ah, it's a TODO.

Reply:
All checks are todo as of now.